src/operator.rs

```diff
@@ -143,12 +143,24 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
                     // somewhat fuzzy about this case, so I think for now this check is
                     // "good enough".
                     // Dead allocations in miri cannot overlap with live allocations, but
-                    // on read hardware this can easily happen. Thus for comparisons we require
+                    // on real hardware this can easily happen. Thus for comparisons we require
                     // both pointers to be live.
-                    self.memory().get(left.alloc_id)?.check_bounds_ptr(left)?;
-                    self.memory().get(right.alloc_id)?.check_bounds_ptr(right)?;
-                    // Two in-bounds pointers, we can compare across allocations
-                    left == right
+
+                    let check = |ptr: Pointer<Borrow>| match self.memory().get(ptr.alloc_id) {
```

Comment:
This is not nice. :/

For once, it's not correct: A pointer into a function allocation is only in-bounds if its offset is 0.

But also, the previous behavior that `memory().check_bounds` would accept offset-0-size-0 pointers into function allocations was fully intended. These pointers *are* in-bounds. There is nothing wrong with a `&()` pointing to a function allocation. So I don't think fixing this here is how we want to proceed.

For example, `pointer_offset_inbounds` now also fails to compute a 0-offset on a function pointer. This affects every user of the old `check_bounds` method that was converted to `memory().get`.

Reply:
I can reintroduce the `check_bounds` method on `Memory`

Reply:
Yeah I think that would make most sense.